src/core/parser.js

```diff
@@ -40,27 +40,20 @@ import { PredictorStream } from "./predictor_stream.js";
 import { RunLengthStream } from "./run_length_stream.js";
 
 const MAX_LENGTH_TO_CACHE = 1000;
-const MAX_ADLER32_LENGTH = 5552;
 
-function computeAdler32(bytes) {
-  const bytesLength = bytes.length;
-  if (
-    typeof PDFJSDev === "undefined" ||
-    PDFJSDev.test("!PRODUCTION || TESTING")
-  ) {
-    assert(
-      bytesLength < MAX_ADLER32_LENGTH,
-      'computeAdler32: Unsupported "bytes" length.'
-    );
-  }
-  let a = 1,
-    b = 0;
-  for (let i = 0; i < bytesLength; ++i) {
-    // No modulo required in the loop if `bytesLength < 5552`.
-    a += bytes[i] & 0xff;
-    b += a;
+function getInlineImageCacheKey(bytes) {
+  const strBuf = [];
+  for (let i = 0, ii = bytes.length; i < ii; i += 2) {
+    if (i === ii - 1) {
```

Comment:
You could move out this check outside of the loop and check if the length is odd or not.
And I think we should add bytes.length.toString into the strBuf to avoid any issues: whatever the way to combine the bytes I always find a way to have identical results with different buffers.

Reply:
I think we could simply all this stuff in something like:
```js
if (bytes.length % 2 === 0) {
  return String.fromCharCode.apply(null, new Uint16Array(bytes.buffer, bytes.byteOffset, bytes.byteLength));
} else {
 return String.fromCharCode.apply(null, new Uint16Array(bytes.buffer, bytes.byteOffset, bytes.byteLength - 1))
   + String.fromCharCode(bytes[bytes.length - 1]);
}
```
Wdyt ?

Reply:
> You could move out this check outside of the loop and check if the length is odd or not.

Yes, that probably wouldn't hurt.

> I think we could simply all this stuff in something like:

Unfortunately that doesn't seem to work, since the `Uint8Array` that we're passing in shares its underlying buffer with the `Stream` that's used to represent the entire /Page.
In order to do what you suggest we'd probably need to copy the `Uint8Array` first, which would thus increase peak memory usage even more here.

Reply:
It's likely because the length argument is related to new array so here the last param should be divided by 2.
So `String.fromCharCode.apply(null, new Uint16Array(bytes.buffer, bytes.byteOffset, bytes.byteLength >> 1))` should give the main string and then just to append the last char.

Reply:
Sorry, forgot to mention the other problem I encountered when trying that previously:
```
Warning: getOperatorList - ignoring errors during "GetOperatorList: page 0" task: "RangeError: start offset of Uint16Array should be a multiple of 2".
```
Given that this is arbitrary `Stream` data, there's thus no guarantees about the starting position of the data.

Reply:
Ah yes it was what I mentioned when talking about aligned data.
So either we keep your code for the general case or we can use the Uint16Array trick in case bytes.byteOffset is a multiple of 2 and else your code.
The Uint16Array trick should be faster.
But I'd understand that you'd prefer to not increase the complexity of the code so it's up to you.

Reply:
> Ah yes it was what I mentioned when talking about aligned data.

Got it, thanks!

> The Uint16Array trick should be faster.

Having just tried it, for issue 2618 it looks like a very significant regression unfortunately: it goes from <3 seconds to >11 seconds in total rendering time.

> But I'd understand that you'd prefer to not increase the complexity of the code so it's up to you.

Yes, I prefer it for both complexity and performance reasons :-)

Reply:
Could you attach the pdf to issue 2618 ?

Reply:
[issue2618.pdf](https://github.com/mozilla/pdf.js/files/9983741/issue2618.pdf)

Reply:
In looking in the profiler, it seems that the code isn't jitted when we've the Uint16Array: I'll file a bug.